test/unittests/test_cpp.cpp

```diff
@@ -10,11 +10,221 @@
+}
+
+TEST(cpp, address_comparison)
+{
+    const auto zero = evmc::address{};
+    for (size_t i = 0; i < sizeof(evmc::address); ++i)
+    {
+        auto t = evmc::address{};
+        t.bytes[i] = 1;
+        auto u = evmc::address{};
+        u.bytes[i] = 2;
+        auto f = evmc::address{};
+        f.bytes[i] = 0xff;
+
+        EXPECT_TRUE(zero < t);
+        EXPECT_TRUE(zero < u);
+        EXPECT_TRUE(zero < f);
+        EXPECT_TRUE(zero != t);
+        EXPECT_TRUE(zero != u);
+        EXPECT_TRUE(zero != f);
+
+        EXPECT_TRUE(t < u);
+        EXPECT_TRUE(t < f);
+        EXPECT_TRUE(u < f);
+
+        EXPECT_FALSE(u < t);
+        EXPECT_FALSE(f < t);
+        EXPECT_FALSE(f < u);
+
+        EXPECT_TRUE(t != u);
+        EXPECT_TRUE(t != f);
+        EXPECT_TRUE(u != t);
+        EXPECT_TRUE(u != f);
+        EXPECT_TRUE(f != t);
+        EXPECT_TRUE(f != u);
+
+        EXPECT_TRUE(t == t);
+        EXPECT_TRUE(u == u);
+        EXPECT_TRUE(f == f);
+    }
+}
+
+TEST(cpp, bytes32_comparison)
+{
+    const auto zero = evmc::bytes32{};
+    for (size_t i = 0; i < sizeof(evmc::bytes32); ++i)
+    {
+        auto t = evmc::bytes32{};
+        t.bytes[i] = 1;
+        auto u = evmc::bytes32{};
+        u.bytes[i] = 2;
+        auto f = evmc::bytes32{};
+        f.bytes[i] = 0xff;
+
+        EXPECT_TRUE(zero < t);
+        EXPECT_TRUE(zero < u);
+        EXPECT_TRUE(zero < f);
+        EXPECT_TRUE(zero != t);
+        EXPECT_TRUE(zero != u);
+        EXPECT_TRUE(zero != f);
+
+        EXPECT_TRUE(t < u);
+        EXPECT_TRUE(t < f);
+        EXPECT_TRUE(u < f);
+
+        EXPECT_FALSE(u < t);
+        EXPECT_FALSE(f < t);
+        EXPECT_FALSE(f < u);
+
+        EXPECT_TRUE(t != u);
+        EXPECT_TRUE(t != f);
+        EXPECT_TRUE(u != t);
+        EXPECT_TRUE(u != f);
+        EXPECT_TRUE(f != t);
+        EXPECT_TRUE(f != u);
+
+        EXPECT_TRUE(t == t);
+        EXPECT_TRUE(u == u);
+        EXPECT_TRUE(f == f);
+    }
+}
+
+TEST(cpp, literals)
+{
+    using namespace evmc::literals;
+
+#if !defined(_MSC_VER) || (_MSC_VER >= 1910 /* Only for Visual Studio 2017+ */)
+    constexpr auto address1 = 0xa0a1a2a3a4a5a6a7a8a9d0d1d2d3d4d5d6d7d8d9_addr;
+    constexpr auto hash1 = 0x01020304050607080910a1a2a3a4a5a6a7a8a9b0c1c2c3c4c5c6c7c8c9d0d1d2_b32;
+    constexpr auto zero_address = 0_addr;
+    constexpr auto zero_hash = 0_b32;
+
+    static_assert(address1.bytes[0] == 0xa0, "");
+    static_assert(address1.bytes[9] == 0xa9, "");
+    static_assert(address1.bytes[10] == 0xd0, "");
+    static_assert(address1.bytes[19] == 0xd9, "");
+    static_assert(hash1.bytes[0] == 0x01, "");
+    static_assert(hash1.bytes[10] == 0xa1, "");
+    static_assert(hash1.bytes[31] == 0xd2, "");
+    static_assert(zero_address == evmc::address{}, "");
+    static_assert(zero_hash == evmc::bytes32{}, "");
+#endif
+
+    EXPECT_EQ(0_addr, evmc::address{});
+    EXPECT_EQ(0_b32, evmc::bytes32{});
+
+    auto a1 = 0xa0a1a2a3a4a5a6a7a8a9d0d1d2d3d4d5d6d7d8d9_addr;
```

Comment:
Since these are long literals anyway, I see no benefit shortening `address` to `addr` and `bytes32` to `b32`.